File: python/skill_extractor/matcher.py
"""Keyword matcher — a faithful port of FlashText's extract_keywords.

Word boundaries are any characters outside [A-Za-z0-9_]. Matching is
case-insensitive, longest-match-wins, non-overlapping. Ported 1:1 from
flashtext.KeywordProcessor so every language implementation of
skill-extractor produces identical spans.
"""

_WORD_CHARS = frozenset(
    "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ_"
)
_KEYWORD = object()  # trie leaf sentinel


class KeywordMatcher:
    def __init__(self, keywords=None):
        self._trie = {}
        for kw in keywords or []:
            self.add(kw)

    def add(self, keyword: str) -> None:
        node = self._trie
        for ch in keyword.lower():
            node = node.setdefault(ch, {})
        node[_KEYWORD] = keyword.lower()

    def extract(self, sentence: str) -> list[tuple[str, int, int]]:
        """Return (keyword, start, end) spans, matching FlashText exactly."""
        out = []
        if not sentence:
            return out
        sentence = sentence.lower()
        current = self._trie
        seq_start = 0
        seq_end = 0
        reset = False
        idx = 0
        n = len(sentence)
        while idx < n:
            ch = sentence[idx]
            if ch not in _WORD_CHARS:
                if _KEYWORD in current or ch in current:
                    longest = None
                    longer_found = False
                    if _KEYWORD in current:
                        longest = current[_KEYWORD]
                        seq_end = idx
                    if ch in current:
                        cont = current[ch]
                        idy = idx + 1
                        broke = False
                        while idy < n:
                            inner = sentence[idy]
                            if inner not in _WORD_CHARS and _KEYWORD in cont:
                                longest = cont[_KEYWORD]
                                seq_end = idy
                                longer_found = True
                            if inner in cont:
                                cont = cont[inner]
                            else:
                                broke = True
                                break
                            idy += 1
                        if not broke:  # reached end of sentence
                            if _KEYWORD in cont:
                                longest = cont[_KEYWORD]
                                seq_end = idy
                                longer_found = True
                        if longer_found:
                            idx = seq_end
                    current = self._trie
                    if longest:
                        out.append((longest, seq_start, idx))
                    reset = True
                else:
                    current = self._trie
                    reset = True
            elif ch in current:
                current = current[ch]
            else:
                current = self._trie
                reset = True
                idy = idx + 1
                while idy < n:
                    if sentence[idy] not in _WORD_CHARS:
                        break
                    idy += 1
                idx = idy
            if idx + 1 >= n:
                if _KEYWORD in current:
                    out.append((current[_KEYWORD], seq_start, n))
            idx += 1
            if reset:
                reset = False
                seq_start = idx
        return out

File: ruby/lib/skill_extractor/matcher.rb
# Keyword matcher — a faithful port of FlashText's extract_keywords.
# Word boundaries: any char outside [A-Za-z0-9_]. Case-insensitive,
# longest-match-wins, non-overlapping. Identical spans across all
# skill-extractor language implementations.
module SkillExtractor
  class KeywordMatcher
    KEYWORD = :__keyword__
    WORD_CHARS = ("a".."z").to_a + ("A".."Z").to_a + ("0".."9").to_a + ["_"]
    WORD_SET = WORD_CHARS.to_h { |c| [c, true] }.freeze

    def initialize(keywords = [])
      @trie = {}
      keywords.each { |kw| add(kw) }
    end

    def add(keyword)
      node = @trie
      kw = keyword.downcase
      kw.each_char do |ch|
        node = (node[ch] ||= {})
      end
      node[KEYWORD] = kw
    end

    # Returns [[keyword, start, end], ...] spans, matching FlashText exactly.
    def extract(sentence)
      out = []
      return out if sentence.nil? || sentence.empty?

      sentence = sentence.downcase
      chars = sentence.chars
      current = @trie
      seq_start = 0
      seq_end = 0
      reset = false
      idx = 0
      n = chars.length
      while idx < n
        ch = chars[idx]
        if !WORD_SET[ch]
          if current.key?(KEYWORD) || current.key?(ch)
            longest = nil
            longer_found = false
            if current.key?(KEYWORD)
              longest = current[KEYWORD]
              seq_end = idx
            end
            if current.key?(ch)
              cont = current[ch]
              idy = idx + 1
              broke = false
              while idy < n
                inner = chars[idy]
                if !WORD_SET[inner] && cont.key?(KEYWORD)
                  longest = cont[KEYWORD]
                  seq_end = idy
                  longer_found = true
                end
                if cont.key?(inner)
                  cont = cont[inner]
                else
                  broke = true
                  break
                end
                idy += 1
              end
              if !broke && cont.key?(KEYWORD)
                longest = cont[KEYWORD]
                seq_end = idy
                longer_found = true
              end
              idx = seq_end if longer_found
            end
            current = @trie
            out << [longest, seq_start, idx] if longest
            reset = true
          else
            current = @trie
            reset = true
          end
        elsif current.key?(ch)
          current = current[ch]
        else
          current = @trie
          reset = true
          idy = idx + 1
          while idy < n
            break unless WORD_SET[chars[idy]]
            idy += 1
          end
          idx = idy
        end
        if idx + 1 >= n && current.key?(KEYWORD)
          out << [current[KEYWORD], seq_start, n]
        end
        idx += 1
        if reset
          reset = false
          seq_start = idx
        end
      end
      out
    end
  end
end
